lint/unusedParam_checker.go

```diff
@@ -0,0 +1,54 @@
+package lint
+
+import (
+	"go/ast"
+)
+
+func init() {
+	addChecker(unusedParamChecker{}, &ruleInfo{})
+}
+
+type unusedParamChecker struct {
+	baseFuncDeclChecker
+}
+
+func (c unusedParamChecker) New(ctx *context) func(*ast.File) {
+	return wrapFuncDeclChecker(&unusedParamChecker{
+		baseFuncDeclChecker: baseFuncDeclChecker{ctx: ctx},
+	})
+}
+
+func (c *unusedParamChecker) CheckFuncDecl(decl *ast.FuncDecl) {
+	params := decl.Type.Params
+	if params == nil || params.NumFields() == 0 {
+		return
+	}
+
+	objUsed := make(map[*ast.Object]struct{})
+	for id := range c.ctx.TypesInfo.Uses {
+		objUsed[id.Obj] = struct{}{}
+	}
```

Comment:
My main concern is that this checker does whole `TypesInfo.Uses` copy for every declaration.
For stdlib package `reflect`, len(TypesInfo.Uses) is 9092 and checker is invoked on 195 functions there.
1772940 map assignments for a single package can be noticeable O(N*N).

I don't have a solution that I like right now, unfortunately.

Reply:
We can invert loop below, so instead of creating a map and iterating over params, we can iterate over params first and check if it was used.

Reply:
If you can make it work, send an updated patch set.
`types.Info.Uses` can hold huge amount of entries for some packages, I remember this being an issue in some of my previous projects, so we definitely want to avoid looping through it more than once.